src/Streamly/System/Process.hs

```diff
@@ -172,56 +169,88 @@ wait procHandle = liftIO $ do
+-- Set everything explicitly so that we are immune to changes upstream.
+procAttrs :: FilePath -> [String] -> CreateProcess
+procAttrs path args = CreateProcess
+    { cmdspec = RawCommand path args
+    , cwd = Nothing -- inherit
+    , env = Nothing -- inherit
+
+    -- File descriptors
+    , std_in = CreatePipe
+    , std_out = CreatePipe
+    , std_err = Inherit
+    , close_fds = True  -- False?
+
+    -- Session/group/setuid/setgid
+    , create_group = False
+    , child_user = Nothing  -- Posix only
+    , child_group = Nothing  -- Posix only
+
+    -- Signals (Posix only) behavior
+    -- Reset SIGINT (Ctrl-C) and SIGQUIT (Ctrl-\) to default handlers.
+    , delegate_ctlc = False
+
+    -- Terminal behavior
+    , new_session = False  -- Posix only
+    , detach_console = False -- Windows only
+    , create_new_console = False -- Windows Only
+
+    -- Added by commit 6b8ffe2ec3d115df9ccc047599545ca55c393005
+    , use_process_jobs = True -- Windows only
+    }
+
+pipeStdErr :: CreateProcess -> CreateProcess
+pipeStdErr cfg = cfg { std_err = CreatePipe }
+
+-- | Creates a system process from an executable path and arguments. For the
+-- default attributes used to create the process see 'procAttrs'.
+--
+createProc' ::
+       (CreateProcess -> CreateProcess) -- ^ Process attribute modifier
+    -> FilePath                         -- ^ Executable path
+    -> [String]                         -- ^ Arguments
     -> IO (Handle, Handle, Handle, ProcessHandle)
     -- ^ (Input Handle, Output Handle, Error Handle, Process Handle)
-openProcErr path args = do
-    let spec = proc path args
-        spec1 =
-            spec
-                { std_in = CreatePipe
-                , std_out = CreatePipe
-                , std_err = CreatePipe
-                , close_fds = True
-                , use_process_jobs = True
-                }
-
+createProc' modCfg path args = do
     (Just stdinH, Just stdoutH, Just stderrH, procH)
-        <- createProcess spec1
+        <- createProcess $ modCfg $ procAttrs path args
     return (stdinH, stdoutH, stderrH, procH)
 
-{-# INLINE processChunks #-}
-processChunks ::
+{-# INLINE runStdin #-}
+runStdin :: (MonadIO m, IsStream t) => t m (Array Word8) -> Handle -> t m a
+runStdin input stdinH =
+    Stream.before
+        (Handle.putChunks stdinH (adapt input) >> liftIO (hClose stdinH))
+        Stream.nil
+
+{-# INLINE processChunksWith #-}
+processChunksWith ::
     (IsStream t, MonadCatch m, MonadAsync m)
-    => FilePath             -- ^ Path to Executable
+    => ((Handle, Handle, Handle, ProcessHandle) -> t m a)
+    -> FilePath             -- ^ Path to Executable
     -> [String]             -- ^ Arguments
-    -> t m (Array Word8)    -- ^ Input stream
-    -> t m (Either (Array Word8) (Array Word8))     -- ^ Output stream
-processChunks path args input = Stream.bracket alloc cleanup run
+    -> t m a     -- ^ Output stream
+processChunksWith run path args = Stream.bracket alloc cleanup run
```

Comment:
`run` generates the input stream?

Reply:
It handles all the streams (stdin, stdout, stderr) in parallel, each stream is handled in a different thread.